src/js/plugin.js

```diff
@@ -0,0 +1,436 @@
+      }
+    });
+
+    // If the player is disposed, dispose the plugin.
+    player.on('dispose', this.dispose);
+
+    /**
+     * Signals that a plugin (both basic and class-based) has just been set up
+     * on a player.
+     *
+     * In all cases, an object containing the following properties is passed as a
+     * second argument to event listeners:
+     *
+     * - `name`: The name of the plugin that was set up.
+     * - `plugin`: The raw plugin function.
+     * - `instance`: For class-based plugins, the instance of the plugin sub-class,
+     *   but, for basic plugins, the return value of the plugin invocation.
+     *
+     * @event pluginsetup
+     * @memberof Player
+     * @instance
+     */
+    player.trigger('pluginsetup', this.getEventHash_());
+  }
+
+  /**
+   * Each event triggered by plugins includes a hash of additional data with
+   * conventional properties.
+   *
+   * This returns that object or mutates an existing hash.
+   *
+   * @private
+   * @param  {Object} [hash={}]
+   *         An object to be used as event an event hash.
+   *
+   * @return {Object}
+   *         The event hash object with, at least, the following properties:
+   *
+   *         - `instance`: The plugin instance on which the event is fired.
+   *         - `name`: The name of the plugin.
+   *         - `plugin`: The plugin class/constructor.
+   */
+  getEventHash_(hash = {}) {
+    hash.name = this.name;
+    hash.plugin = this.constructor;
+    hash.instance = this;
+    return hash;
+  }
+
+  /**
+   * Triggers an event on the plugin object.
+   *
+   * @param  {Event|Object|string} event
+   *         A string (the type) or an event object with a type attribute.
+   *
+   * @param  {Object} [hash={}]
+   *         Additional data hash to pass along with the event. For plugins,
+   *         several properties are added to the hash:
+   *
+   *         - `instance`: The plugin instance on which the event is fired.
+   *         - `name`: The name of the plugin.
+   *         - `plugin`: The plugin class/constructor.
+   *
+   * @return {boolean}
+   *         Whether or not default was prevented.
+   */
+  trigger(event, hash = {}) {
+    return Events.trigger(this.eventBusEl_, event, this.getEventHash_(hash));
+  }
+
+  /**
+   * Handles "statechanged" events on the plugin. No-op by default, override by
+   * subclassing.
+   *
+   * @param {Event} e
+   *        An event object provided by a "statechanged" event.
+   *
+   * @param {Object} e.changes
+   *        An object describing changes that occurred with the "statechanged"
+   *        event.
+   */
+  handleStateChanged(e) {}
+
+  /**
+   * Disposes a plugin.
+   *
+   * Subclasses can override this if they want, but for the sake of safety,
+   * it's probably best to subscribe the "dispose" event.
+   */
+  dispose() {
+    const {name, player} = this;
+
+    /**
+     * Signals that a class-based plugin is about to be disposed.
+     *
+     * @event dispose
```

Comment:
* `@event Plugin#dispose` 
* remove `@memberof` & `@instance`
* add in `@type {EventTarget~Event}` as it is just a standard event